### PR TITLE
Convert startProductRequest(), finishTransaction(), restoreTransactions(), presentCodeRedemptionSheet() to pigeon

### PR DESCRIPTION
**Pull request by LouiseHsu**
Part 2 of https://github.com/flutter/flutter/issues/117910

This PR converts startProductRequest(), finishTransaction(), restoreTransactions(), presentCodeRedemptionSheet() to pigeon, as well as add all remaining converts to and from pigeons for SK objects.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/Runner.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -666,7 +666,7 @@
 				CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF = YES;
 				CLANG_WARN_UNGUARDED_AVAILABILITY = YES_AGGRESSIVE;
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = S8QB4VV633;
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/example/ios/Runner.xcodeproj/project.pbxproj:669]: check out this part

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m [hunk 2]
```diff
@@ -147,38 +136,34 @@ - (nullable SKStorefrontMessage *)storefrontWithError:(FlutterError *_Nullable *
...
+      error = [FlutterError errorWithCode:@"storekit_getproductrequest_platform_error"
+                                  message:startProductRequestError.localizedDescription
+                                  details:startProductRequestError.description];
     }
     if (!response) {
-      result([FlutterError errorWithCode:@"storekit_platform_no_response"
-                                 message:@"Failed to get SKProductResponse in startRequest "
-                                         @"call. Error occured on iOS platform"
-                                 details:call.arguments]);
-      return;
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m:176]: same here

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m [hunk 3]
```diff
@@ -265,35 +244,27 @@ - (void)finishTransaction:(FlutterMethodCall *)call result:(FlutterResult)result
...
+                                                   error {
+  [self.paymentQueueHandler restoreTransactions:applicationUserName];
 }
 
+- (void)presentCodeRedemptionSheetWithError:
+    (FlutterError *_Nullable __autoreleasing *_Nonnull)error {
 #if TARGET_OS_IOS
-- (void)presentCodeRedemptionSheet:(FlutterMethodCall *)call result:(FlutterResult)result {
   [self.paymentQueueHandler presentCodeRedemptionSheet];
-  result(nil);
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m:294]: I just want to call out how these calls have disappears now.  In the past, failing to call these created bugs, you've eliminated that risk now 👍

File: packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/InAppPurchasePluginTests.m [hunk 4]
```diff
@@ -105,22 +105,17 @@ - (void)testPaymentQueueStorefrontReturnsNil {
...
-                         }];
-  [self waitForExpectations:@[ expectation ] timeout:5];
-  XCTAssert([result isKindOfClass:[NSDictionary class]]);
-  NSArray *resultArray = [result objectForKey:@"products"];
-  XCTAssertEqual(resultArray.count, 1);
-  XCTAssertTrue([resultArray.firstObject[@"productIdentifier"] isEqualToString:@"123"]);
+  NSArray *argument = @[ @"123" ];
+  [self.plugin
+      startProductRequestProductIdentifiers:argument
+                                 completion:^(SKProductsResponseMessage *_Nullable response,
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/InAppPurchasePluginTests.m:111]: Can you add an assertion that the completion block was called too?  Here's an example of that: https://github.com/flutter/packages/pull/6069/files#diff-3033f32996418c96cac1b73875958073e57facf79ffdbc7e6864f8b27a63de37R21

File: packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/TranslatorTests.m [hunk 5]
```diff
@@ -458,4 +458,40 @@ - (void)testSKPaymentTransactionConvertToPigeon {
...
+      [[SKProductsResponseStub alloc] initWithMap:self.productResponseMap];
+  SKProductsResponseMessage *responseMsg =
+      [FIAObjectTranslator convertProductsResponseToPigeon:response];
+
+  XCTAssertEqual(responseMsg.products.count, 1);
+  XCTAssertEqual(responseMsg.invalidProductIdentifiers.count, 0);
+
+  SKProductMessage *productMsg = responseMsg.products[0];
+
+  XCTAssertEqualObjects(productMsg.price, @"1");
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/example/shared/RunnerTests/TranslatorTests.m:472]: nit: all of these asserts are against values that are being set in `setUp` right? It's kind of far from here.  Adding a comment wouldn't hurt.

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart [hunk 6]
```diff
@@ -164,7 +242,39 @@ enum SKProductDiscountPaymentMode {
...
+        return SKProductDiscountPaymentMode.payUpFront;
+      case SKProductDiscountPaymentModeMessage.freeTrial:
+        return SKProductDiscountPaymentMode.freeTrail;
+      case SKProductDiscountPaymentModeMessage.unspecified:
+        return SKProductDiscountPaymentMode.unspecified;
+    }
+  }
+
+  @visibleForTesting
+
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart:263]: Remove newline please

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart [hunk 7]
```diff
@@ -182,7 +292,31 @@ enum SKProductDiscountType {
...
+    switch (msg) {
+      case SKProductDiscountTypeMessage.introductory:
+        return SKProductDiscountType.introductory;
+      case SKProductDiscountTypeMessage.subscription:
+        return SKProductDiscountType.subscription;
+    }
+  }
+
+  @visibleForTesting
+
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart:309]: same

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart [hunk 8]
```diff
@@ -383,6 +550,53 @@ class SKProductWrapper {
       subscriptionPeriod,
       introductoryPrice,
       discounts);
+
+  /// Convert from [SKProductMessage] to [SKProductWrapper]
+  static SKProductWrapper convertFromPigeon(SKProductMessage msg) {
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart:555]: No action required: These are all a pain.  I guess if pigeon had been used from day one, the generated classes could have been been used.  We also talked about letting pigeon use existing classes too.  That would have really helped here.

File: packages/in_app_purchase/in_app_purchase_storekit/pigeons/messages.dart [hunk 9]
```diff
@@ -130,6 +130,112 @@ class SKStorefrontMessage {
...
+  SKProductSubscriptionPeriodMessage(
+      {required this.numberOfUnits, required this.unit});
+
+  final int numberOfUnits;
+  final SKSubscriptionPeriodUnitMessage unit;
+}
+
+enum SKSubscriptionPeriodUnitMessage {
+  day,
+
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/pigeons/messages.dart:231]: remove newlines please

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart [hunk 10]
```diff
@@ -288,4 +195,57 @@ class FakeStoreKitPlatform implements TestInAppPurchaseApi {
...
+  @override
+  void finishTransaction(Map<String?, String?> finishMap) {
+    finishedTransactions.add(createPurchasedTransaction(
+        finishMap['productIdentifier']!, finishMap['transactionIdentifier']!,
+        quantity: transactionList.first.payment.quantity));
+  }
+
+  @override
+  void presentCodeRedemptionSheet() {
+    return;
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart:208]:
```suggestion
```

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart [hunk 11]
```diff
@@ -383,6 +550,53 @@ class SKProductWrapper {
...
+        discounts: msg.discounts != null
+            ? msg.discounts!
+                .map((SKProductDiscountMessage? e) =>
+                    SKProductDiscountWrapper.convertFromPigeon(e!))
+                .toList()
+            : <SKProductDiscountWrapper>[]);
+  }
+
+  /// Convert from [SKProductWrapper] to [SKProductMessage]
+  static SKProductMessage convertToPigeon(SKProductWrapper wrapper) {
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_product_wrapper.dart:579]: Do all of these have tests?

**Review by gaaclarke — COMMENTED** [review]
Looking good, the only major thing is the missing short circuiting and potentially missing tests for all these extra converters.

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/FIAObjectTranslator.m [hunk 12]
```diff
@@ -470,7 +470,7 @@ + (nullable SKProductMessage *)convertProductToPigeon:(nullable SKProduct *)prod
 
   NSArray<SKProductDiscount *> *skProductDiscounts = product.discounts;
   NSMutableArray<SKProductDiscountMessage *> *pigeonProductDiscounts =
-      [[NSMutableArray alloc] init];
+      [[NSMutableArray arrayWithCapacity:skProductDiscounts.count] init];
```
Comment (gaaclarke) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/FIAObjectTranslator.m:473]: You don't want to call `init` when using `arrayWithCapacity:`, that's already taken care for you.  You just need that when you are using `alloc`.

Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/FIAObjectTranslator.m:473]: oh oopsies

**Review by gaaclarke — APPROVED** [review]
LGTM!  Don't forget to manually test it with the example app before merging.

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m [hunk 13]
```diff
@@ -265,35 +248,27 @@ - (void)finishTransaction:(FlutterMethodCall *)call result:(FlutterResult)result
       @try {
         [self.paymentQueueHandler finishTransaction:transaction];
       } @catch (NSException *e) {
-        result([FlutterError errorWithCode:@"storekit_finish_transaction_exception"
-                                   message:e.name
-                                   details:e.description]);
+        *error = [FlutterError errorWithCode:@"storekit_finish_transaction_exception"
+                                     message:e.name
+                                     details:e.description];
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m:253]: where is this error used?

**Review by hellohuanlin — COMMENTED** [review]
just briefly skimmed through since Aaron already reviewed it

**Comment by gaaclarke** [comment]
I'm not sure why that CI step keeps failing, it's an infra failure so it's likely not caused by this PR.  You can try updating your branch by merging in main or rebasing against main.

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m [hunk 14]
```diff
@@ -88,18 +88,7 @@ - (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar
...
-    [self finishTransaction:call result:result];
-  } else if ([@"-[InAppPurchasePlugin restoreTransactions:result:]" isEqualToString:call.method]) {
-    [self restoreTransactions:call result:result];
-#if TARGET_OS_IOS
-  } else if ([@"-[InAppPurchasePlugin presentCodeRedemptionSheet:result:]"
-                 isEqualToString:call.method]) {
-    [self presentCodeRedemptionSheet:call result:result];
-#endif
-  } else if ([@"-[InAppPurchasePlugin retrieveReceiptData:result:]" isEqualToString:call.method]) {
+  if ([@"-[InAppPurchasePlugin retrieveReceiptData:result:]" isEqualToString:call.method]) {
```
[the remark this answers is not in the document]
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m:91]: this entire code block/fcn will be removed after the pigeon migration so it should be fine

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m (hunk 13 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.m:253]: all generated pigeon functions have like an extra "error" param, I think it gets passed back to the dart side automatically

**Pull request merged**